### ISSUE TITLE
Investigate Erroneous GPS positions

### ISSUE THREAD

**Issue by radumas** [CityofToronto/bdit_data-sources#96]
Per investigations into the CIS Data, there appear to be patterns in erroneous GPS positions:

 - A band across the southern part of the map
 - what appear to be routes, at a much larger scale, to the North of the City

![image 1](https://user-images.githubusercontent.com/10623872/37790031-cdcdc768-2ddb-11e8-9569-df5faf66e9e4.png)

From a comparison with previously shared NVAS data (yellow lines), we can infer a few observations:

 - This also affects the data output in realtime through the NVAS feed
 - From a quick glance at the lines, the outlying data points appear to be unique extreme values that somehow form patterns.

![image 2](https://user-images.githubusercontent.com/10623872/37790171-2476266e-2ddc-11e8-8954-36252d81dfe2.png)

 - [ ] Investigate outlying points further, are there any patterns?
 - [x] Investigate the band of points further, are there any patterns?

**Comment by radumas** [comment]
For October 4th, 2017 there are 680,330 CIS observations for 15 streetcar routes, with 306 vehicles, the distribution by vehicle series is:

veh_series|count
|----------|------:|
'1XXX'|52
'4XXX'|174
'7XXX'|30
'8XXX'|50

Focusing on 504/514 first, there are 136,363 observations:
```sql
SELECT LEFT(vehicle::TEXT, 1) || 'XXX' AS veh_series,
 COUNT(DISTINCT vehicle) AS "Num vehicles",  COUNT(1) AS "Num points"
  FROM ttc.cis_20171004
  WHERE route IN ('504','514')
  GROUP BY veh_series   ORDER BY veh_series
```

'veh_series'|'Num vehicles'|'Num points'
|-----------|----------------:|--------------:|
'4XXX'|72|129625
'7XXX'|4|3317
'8XXX'|4|3421

The below map shows the downtown GPS points for the day. Purple lines are overlayed representing the scheduled GPS route shapes. The mustard polygon is approximately the "Downtown" for purposes of comparing GPS points along those routes vs. outlier values. Some quick observations:

 - Vehicles appear to interline with other routes after Broadview (street and station). **This means we might be missing data not identified being on a streetcar route**
 - GPS values get very noisy in the King Street Pilot area.
 - the highlighted straight line does not appear to follow an actual street.

![image](https://user-images.githubusercontent.com/10623872/37851842-cf3db442-2eb6-11e8-94ea-1163674d5bcc.png)

The following query identifies the proportion of vehicles and points within the mustard shape 

```sql
SELECT CASE WHEN vehicle BETWEEN 4000 AND 4200 THEN 'CLRV'
	WHEN vehicle BETWEEN 4200 AND 4300 THEN 'ALRV'
	WHEN vehicle BETWEEN 4400 AND 4500 THEN 'Flexity Freedom'
	ELSE 'Bus Series '|| LEFT(vehicle::TEXT, 2) || 'XX'
	END AS "Vehicle Series",
COUNT(DISTINCT (CASE WHEN s.geom IS NOT NULL THEN vehicle ELSE NULL END) ) AS "Vehicles within DT shape", 
 COUNT(DISTINCT vehicle) AS "Total Vehicles", 
to_char(100.0*(SUM(CASE WHEN s.geom IS NOT NULL THEN 1 ELSE 0 END))/ COUNT(1), '99.9')
 AS "% Points Downtown", 
 COUNT(1) AS "Num points"
  FROM ttc.cis_20171004 c
  LEFT OUTER JOIN rdumas.downtown_streetcars s ON ST_Within("position", s.geom)
  WHERE route IN ('504','514')
  GROUP BY "Vehicle Series"   ORDER BY "Vehicle Series"
```
Vehicle Series|Vehicles within DT shape|Total Vehicles|% Points Downtown|Num points
|---------------|----------------------------:|----------------:|-----------------------:|-------------:|
ALRV|8|8| 98.3|10177
Bus Series 75XX|1|1| 99.6|788
Bus Series 76XX|1|1| 99.6|958
Bus Series 77XX|1|1| 98.1|738
Bus Series 78XX|1|1| 99.2|833
Bus Series 81XX|1|1| 99.4|776
Bus Series 86XX|1|1| 99.8|832
Bus Series 87XX|2|2| 99.6|1813
CLRV|49|49| 98.9|95740
Flexity Freedom|15|15| 97.8|23708

Ironically, the new vehicles seem to have buggier GPS (by 1.1 %). Or is it one vehicle in particular?

```sql
SELECT vehicle AS "Vehicle ID",
to_char(100.0*(SUM(CASE WHEN s.geom IS NOT NULL THEN 1 ELSE 0 END))/ COUNT(1), '99.9')
 AS "% Points Downtown", 
 COUNT(1) AS "Num points"
  FROM ttc.cis_20171004 c
  LEFT OUTER JOIN rdumas.downtown_streetcars s ON ST_Within("position", s.geom)
  WHERE route IN ('504','514') AND vehicle BETWEEN 4400 AND 4500
  GROUP BY "Vehicle ID"   ORDER BY "Vehicle ID"
```
Vehicle ID|% Points Downtown|Num points
|---------|----------------:|---------:|
4400| 99.5|1507
4403| 99.3|891
4404| 99.3|2307
4409| 99.5|2415
4413| 99.7|1126
4416| 99.6|1570
4418| 99.5|2242
4419| 99.7|2940
4420| 99.5|595
4425| 99.6|2216
4437| 99.6|232
4438| 99.4|2685
4440| 98.6|362
4442| 99.5|2103
**4443**|** 21.1**|**517**

**It's one vehicle in particular**

**Comment by radumas** [comment]
It turns out streetcar 4443's 517 GPS points are all in 5 locations, with 400 of them in the top right corner... 
![image](https://user-images.githubusercontent.com/10623872/37975974-6ad23c9e-31ae-11e8-9dfb-c81562a90642.png)

**Comment by radumas** [comment]
What about points in the horizontal band outside of downtown? After drawing in a rough shape, the following query produced the table below of points in that band for Oct 4th.
```sql
SELECT CASE WHEN vehicle BETWEEN 4000 AND 4200 THEN 'CLRV'
	WHEN vehicle BETWEEN 4200 AND 4300 THEN 'ALRV'
	WHEN vehicle BETWEEN 4400 AND 4500 THEN 'Flexity Freedom'
	ELSE 'Bus Series '|| LEFT(vehicle::TEXT, 2) || 'XX'
	END AS "Vehicle Series",
to_char(100.0*(SUM(CASE WHEN s.geom IS NOT NULL THEN 1 ELSE 0 END))/ COUNT(1), '90.9')||'%'
 AS "% Points in the band", 
 COUNT(1) AS "Num points"
  FROM ttc.cis_20171004 c
  LEFT OUTER JOIN rdumas.cis_explore_shapes s ON ST_Within("position", s.geom) AND shape_name = 'milky_way'
  WHERE route IN ('504','514') 
  GROUP BY "Vehicle Series"   ORDER BY "Vehicle Series"
```
Vehicle Series|% Points in the band|Num points
|---------|----------------:|---------:|
ALRV|  1.6%|10177
Bus Series 75XX|  0.4%|788
Bus Series 76XX|  0.3%|958
Bus Series 77XX|  0.5%|738
Bus Series 78XX|  0.5%|833
Bus Series 81XX|  0.4%|776
Bus Series 86XX|  0.2%|832
Bus Series 87XX|  0.3%|1813
CLRV|  0.9%|95740
Flexity Freedom|  0.4%|23708

The Articulated LRVs seem the more problematic, and the CLRVs have the second highest percentage of points in this band. 

### Are these a few faulty vehicles?
The following query finds the 10 vehicles with the highest proportion of points in the Band.
```sql
  SELECT Vehicle, CASE WHEN vehicle BETWEEN 4000 AND 4200 THEN 'CLRV'
	WHEN vehicle BETWEEN 4200 AND 4300 THEN 'ALRV'
	WHEN vehicle BETWEEN 4400 AND 4500 THEN 'Flexity Freedom'
	ELSE 'Bus Series '|| LEFT(vehicle::TEXT, 2) || 'XX'
	END AS "Vehicle Series",
to_char(100.0*(SUM(CASE WHEN s.geom IS NOT NULL THEN 1 ELSE 0 END))/ COUNT(1), '90.9')||'%'
 AS "% Points in the band", 
 COUNT(1) AS "Num points"
  FROM ttc.cis_20171004 c
  LEFT OUTER JOIN rdumas.cis_explore_shapes s ON ST_Within("position", s.geom) AND shape_name = 'milky_way'
  WHERE route IN ('504','514') 
  GROUP BY vehicle, "Vehicle Series"  ORDER BY 1.0*SUM(CASE WHEN s.geom IS NOT NULL THEN 1 ELSE 0 END)/count(1) DESC LIMIT 10
```
vehicle|Vehicle Series|% Points in the band|Num points
|------|---------|----------------:|---------:|
4081|CLRV| 16.1%|2650
4202|ALRV|  7.5%|1607
4096|CLRV|  6.4%|611
4228|ALRV|  5.3%|412
4041|CLRV|  4.6%|436
4440|Flexity Freedom|  1.4%|362
4078|CLRV|  0.8%|390
4051|CLRV|  0.7%|289
4069|CLRV|  0.7%|460
4155|CLRV|  0.6%|3030

So there are about 6 vehicles who seem to have an above average proportion of points within this band, and then after that the proportions of points in the band are about average.

**Comment by radumas** [comment]
Taking a look at vehicle 4081 to try to see if there are some patterns in this data, I merged each point with its subsequent point to create a line between each pair of observations. This allows one to see which point preceded an erroneous one and potentially illuminate some patterns in the data. Additionally I removed the border on the points and added transparency to highlight point density. The results highlight puzzling patterns.

```sql
SELECT row_number() OVER two_point as order_id, lead(message_datetime) OVER two_point AS next_timestamp, st_linefrommultipoint(st_collect(position, COALESCE(lead(position) OVER two_point, position))) as geom
FROM ttc.cis_20171004
WHERE vehicle = 4081
WINDOW two_point AS ( ORDER BY message_datetime)
```

![image](https://user-images.githubusercontent.com/10623872/38324894-831f69ae-380f-11e8-8e55-bce0f65ee4e3.png)

There are few things to note with the points regularly spaced at the bottom of the map above:

 1. The "Identify Results" on the right are for the point highlighted in red. As the lines connecting this point indicate, the vehicle records this erroneous position over 10 times in the day, with the **exact same** coordinates. 
2. As the lines indicate, these positions are being recorded from previous positions (these are multiple and varied) which appear to be within the area where we might accept streetcars to be travelling.
3. However, also as the lines indicate, the streetcar also appears to occasionally "travel" between these regularly appearing erroneous points. 
4. These regular erroneous points have the **exact same** latitude.
5. Creating the same map with vehicle 4202, an ALRV, reveals a similar pattern with the **exact same points**
6.
